staging/src/k8s.io/client-go/util/workqueue/queue.go

```diff
@@ -276,8 +276,10 @@ func (q *Typed[T]) Get() (item T, shutdown bool) {
 	for q.queue.Len() == 0 && !q.shuttingDown {
 		q.cond.Wait()
 	}
-	if q.queue.Len() == 0 {
-		// We must be shutting down.
+
+	// If queue is shutting exit immediately if we are not draining or
```

Comment:
this seems now to contradict the current doc

> // Get blocks until it can return an item to be processed. If shutdown = true,
> // the caller should end their goroutine. You must call Done with item when you
> // have finished processing it.

Reply:
Good point, I amended the doc

Reply:
A change like this will be contentious, since we are changing the contract of an stable API

Reply:
True but as it stands neither the docs nor implementation are correct.

The docs say for `.ShutDown` that workers are instructed to immediately shut down:
https://github.com/kubernetes/kubernetes/blob/195803cde570ad1025a78e36cdbef76bddbc4c33/staging/src/k8s.io/client-go/util/workqueue/queue.go#L312-L313

However that is not the case - `.Get` will only instruct workers to shutdown when the queue is drained, regardless of `.drain`:
https://github.com/kubernetes/kubernetes/blob/195803cde570ad1025a78e36cdbef76bddbc4c33/staging/src/k8s.io/client-go/util/workqueue/queue.go#L276-L277

Imho either the docs or the behaviour would have to be changed.

I'd be fine with just amending the docs, but I do believe that `.Get` should properly honour `.drain`.

Reply:
it looks like https://github.com/kubernetes/kubernetes/pull/101928/files#diff-b80cf8b3dd694a3e7e152b9980be188ae9219c735bfa2d15f498bdc449a6698bL185 introduced ShutdownWithDrain() and changed the godoc on Shutdown() without actually making Get() behavior change at all with respect to drain.

That commit broke the docs for Shutdown() in 1.23. We shouldn't break the behavior for Shutdown() now as well.

I think we should just change the docs to reflect reality:
* ShutDown(): no new items are added, workers drain existing items using Get(), Shutdown() returns immediately without waiting for drain
* ShutDownWithDrain(): same as ShutDown but waits until items are drained to return

Reply:
/cc @deads2k 
for visibility

Reply:
@liggitt @aojea Please re-review, I've updated the PR to change the docs to reflect the implementation.